datafusion/src/optimizer/hash_build_probe_order.rs

```diff
@@ -187,6 +187,7 @@ impl OptimizerRule for HashBuildProbeOrder {
             | LogicalPlan::CreateExternalTable { .. }
             | LogicalPlan::Explain { .. }
             | LogicalPlan::Union { .. }
+            | LogicalPlan::Join { .. }
```

Comment:
should you also possibly conditionalize this on being a SemiJoin? like

```suggestion
            | LogicalPlan::Join if *join_type == JoinType::Semi { .. }
```

So if we add a new join type in the future it doesn't accidentally get allowed / disallowed in some confusing way

Reply:
Tried that in earlier commit... But than the exhaustiveness checking gets in the way unfortunately.

Reply:
I think the only way the compiler can do it now is by explicitly matching on all the join types, but as this feature is not added https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md that would be quite verbose currently

Reply:
I rewrote it a bit to use a function in the body